File: dynamic_programming/highest_route/HighestRouteValue.java
/*
 * Check README to find the description.
 * <br/>
 * Each step we can move down or right. So starting from the top left cell
 * we're moving line horizontal and line vertical and calculate max
 * value for each cell. Storing previous result in the same array
 * and repeating this until reach the last element of an array.
 */
public class HighestRouteValue {
    public int findMaxRoute(int[][] route) {
        return findMaxRoute(route, 0, 0, route.length - 1, 
                route[0].length - 1);
    }
    
    private int findMaxRoute(int[][] route, int x, int y, int n, int m) {
        for(int i = x; i <= n; i++) route[i][y] += getMax(route, i, y);
        for(int j = y + 1; j <= m; j++) route[x][j] += getMax(route, x, j);
        if(x == n || y == m) return route[n][m];
        return findMaxRoute(route, x + 1, y + 1, n, m);
    }

    private int getMax(int[][] route, int x, int y) {
        if(x == 0 && y == 0) return 0;
        if(x == 0) return route[x][y - 1];
        if(y == 0) return route[x - 1][y];
        return Math.max(route[x - 1][y], route[x][y - 1]);
    }

    private static String toString(int[][] array) {
        int n = array.length - 1;
        StringBuilder res = new StringBuilder("{");
        for(int i = 0; i <= n; i++)
            res.append(java.util.Arrays.toString(array[i])).append("\n");
        res.append("}");
        return res.toString();
    }

    public static void main(String[] args) {
        int[][] route = new int[][] {{1, 2, 3}, {6, 9, 11}, 
            {8, 10, 7}, {4, 9, 5}, {10, 8, 3}};
        System.out.println("Array: \n" + toString(route));
        System.out.println("Max apples: " + 
                new HighestRouteValue().findMaxRoute(route));
    }
}

File: puzzles/highest_route/HighestRouteValue.kt
/**
 * Check README to find description.
 */
class HighestRouteValue {
    fun calculate(cell: Coordinates): Int {
        return nextStep(cell)
    }

    fun nextStep(cell: Coordinates): Int {
        if (cell.n == 0 && cell.m == 0) return matrix[0][0]
        if (cell.n > 0 && cell.m > 0)
            return matrix[cell.n][cell.m] + Math.max(
                    nextStep(Coordinates(cell.n - 1, cell.m)),
                    nextStep(Coordinates(cell.n, cell.m - 1)))
        if (cell.n > 0)
            return matrix[cell.n][cell.m] + nextStep(
                    Coordinates(cell.n - 1, cell.m))
        return matrix[cell.n][cell.m] + nextStep(
                Coordinates(cell.n, cell.m - 1))
    }

    class Coordinates constructor(val n: Int, val m: Int) {

        override fun toString(): String {
            return "{$n, $m}"
        }

        override fun hashCode(): Int {
            return Integer.hashCode((n + 1) * (m + 1) + m)
        }

        override fun equals(other: Any?): Boolean {
            if (other !is Coordinates) {
                return false
            }
            return other.n == n && other.m == m
        }
    }
}

private val matrix = arrayOf(
        intArrayOf(1, 2, 3), 
        intArrayOf(6, 9, 11), 
        intArrayOf(8, 10, 7), 
        intArrayOf(4, 9, 5), 
        intArrayOf(10, 8, 3))
private val N = 5
private val M = 3

fun printMatrix(matrix: Array<IntArray>) {
    for (n in matrix.indices) {
        for (m in 0..matrix[n].size - 1) {
            System.out.print(matrix[n][m].toString() + "\t")
        }
        System.out.println("")
    }
}

fun main(args: Array<String>) {
    printMatrix(matrix)
    System.out.println()
    val value = HighestRouteValue().calculate(
            HighestRouteValue.Coordinates(N - 1, M - 1))
    System.out.println(Integer.toString(value))
}
